### PR TITLE
Implement resolution cache with stripped urls as key.

### PR DESCRIPTION
**Pull request by nmoinvaz**

### CODE REVIEW

File: resolver_cache.c [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+    // Use url without a path or query as the key because most systems implement url path stripping for proxies.
+    char *root_url = get_url_without_path_or_query(url);
+    if (!root_url)
+        return NULL;
+
+    mutex_lock(g_proxy_resolver_cache.results_lock);
+    resolved_s *cache;
+    for (cache = g_proxy_resolver_cache.results; cache; cache = cache->next) {
+        if (strcmp(cache->root_url, root_url) == 0)
```
Comment (sergio-nsk) [inline resolver_cache.c:76]: Should not `stricmp`/`strcasecmp` be used?

File: resolver_cache.c [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    if (cache)
+        return cache->list;
+
+    return NULL;
+}
+
+bool proxy_resolver_cache_init(void) {
+    memset(&g_proxy_resolver_cache, 0, sizeof(g_proxy_resolver_cache));
+    g_proxy_resolver_cache.results_lock = mutex_create();
+    if (!g_proxy_resolver_cache.results_lock)
```
Comment (sergio-nsk) [inline resolver_cache.c:91]: You made `return g_proxy_resolver_cache.results_lock;` w/o if recently.

File: resolver.c [hunk 3]
```diff
@@ -107,8 +126,19 @@ bool proxy_resolver_wait(void *ctx, int32_t timeout_ms) {
...
+        proxy_resolver->listp = proxy_resolver->cached_list;
+        return true;
+    }
+
     if (g_proxy_resolver.proxy_resolver_i->wait(proxy_resolver->base, timeout_ms)) {
-        proxy_resolver->listp = (char *)g_proxy_resolver.proxy_resolver_i->get_list(proxy_resolver->base);
+        proxy_resolver->listp = g_proxy_resolver.proxy_resolver_i->get_list(proxy_resolver->base);
+
+        // Store result in cache if it doesn't exist
+        if (proxy_resolver->use_cache && !proxy_resolver->cached_list)
```
[the remark this answers is not in the document]
Comment (nmoinvaz) [reply resolver.c:139]: I think it should be checking `listp` instead of `cached_list`.

File: resolver_cache.c (hunk 2 above)
Comment (nmoinvaz) [reply resolver_cache.c:91]: I was trying to avoid implict cast to bool from pointer.

File: resolver_cache.c (hunk 1 above)
Comment (nmoinvaz) [reply resolver_cache.c:76]: Fixed.

File: resolver.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (nmoinvaz) [reply resolver.c]: Actually, I have just removed it. I don't use `assert` in this project.

**Review by sergio-nsk — APPROVED** [review]

**Comment by nmoinvaz** [comment]
I'm not sure if I will merge this or not. You brought up a good point about the expiration of cache items which I don't have a good idea about now how to handle.

**Comment by sergio-nsk** [comment]
Chromium caches PAC, you can start from caching it.

**Comment by sergio-nsk** [comment]
BTW Chromium caches the system global proxy settings on browser start, except macOS. `proxyres` can do it also.

**Pull request closed without merging**